mtd: spi-nor: Make support for BoHong bh25q128as generic

themaverickdm opened the pull request:
Move the patches for BoHong bh25q128as out of ramips to make them generic.

Functionality was verified on a Wavlink WL-WN586X3 Rev.a with both main kernel (5.15.y) and testing kernel (6.1.y).

themaverickdm commented:
This is the follow-up PR to https://github.com/openwrt/openwrt/pull/14634 for WL-WN586X3 support.

robimarko commented:
This needs to be sent upstream, there is no point keeping this downstream.

themaverickdm commented:
> This needs to be sent upstream, there is no point keeping this downstream.

I am new to this. Which upstream repo should this be sent to?

robimarko commented:
Hi,

This needs to be sent to the Linux kernel MTD group.
Do note that it is likely that the current code is not up to the current kernel API-s and formatting requirements so it would require a certain amount of rework.

I am also aware that you are not the original author and need this for device support so I hope others chime in as well.

themaverickdm commented:
> Hi,
> 
> This needs to be sent to the Linux kernel MTD group. Do note that it is likely that the current code is not up to the current kernel API-s and formatting requirements so it would require a certain amount of rework.
> 
> I am also aware that you are not the original author and need this for device support so I hope others chime in as well.

Hello, thanks for your response.

So in general, am I assuming correctly that we can keep these patches in pending-* directories while we work to get the needed patches in the mainline kernel?

And thanks, I was looking for which Linux kernel staging repo should I direct the patches to.

robimarko commented:
@Ansuel Your opinion?
I am inclined to merge this since its been hanging around for multiple years though not idea l to keep it downstream

robimarko commented:
Oh there is a public datasheet and it looks like a bog standard SPI-NOR
https://www.e-interlink.com.tw/userUpload/files/BH25Q128AS_v1_0.pdf

Ansuel commented:
I sent the patch upstream.

themaverickdm commented:
Sorry for the delayed response. @Ansuel, thank you for sending the patch upstream. Would you please send me the link to those patches, since I would like to see how it's done?
You mentioned that these will move into backports, correct?
What will happen to this PR, BTW?

robimarko commented:
@themaverickdm Any chance you can dump the SFDP from a running device?
Instructions are here:
https://docs.kernel.org/driver-api/mtd/spi-nor.html#minimum-testing-requirements

themaverickdm commented:
@robimarko, I am attaching a file with the information you asked for. Please let me know if this suffices.
Note that I did not have access to xxd, nor mtd-utils. However, openwrt's "mtd write" and eventual reads while booting up are working fine, both on 5.15.y and 6.1.y kernels, so I would think that it is a good enough test of the working of this device.
[bohong-bh25q128as-info.log](https://github.com/openwrt/openwrt/files/14335239/bohong-bh25q128as-info.log)

Ansuel commented:
> Note that I did not have access to xxd, nor mtd-utils.

just install the packages. Is it a problem?

themaverickdm commented:
Well, this is from my local build, so I will need to figure out how to install packages from local build (haven't done that before).

xabolcs commented:
`hexdump -C` should be installed and provides similar output

themaverickdm commented:
> `hexdump -C` should provide similar output and it's installed by default.

hexdump output is already included in the attached file.

robimarko commented:
@Ansuel Everything Miquel asked from you is in the log, that is enough, it even has an SFDP table

Ansuel commented:
> > `hexdump -C` should provide similar output and it's installed by default.
> 
> hexdump output is already included in the attached file. All, please go through the attached file while I try to figure out how to build and install xxd and mtd-utils from my local build. :)

can't you just select them from make menuconfig?

themaverickdm commented:
> > > `hexdump -C` should provide similar output and it's installed by default.
> > 
> > 
> > hexdump output is already included in the attached file. All, please go through the attached file while I try to figure out how to build and install xxd and mtd-utils from my local build. :)
> 
> can't you just select them from make menuconfig?

xxd output included in this newly attached file. Please take a look. mtd-utils can really only be run via initramfs, right, since there's no extra space on this device to test it on?

[bohong-bh25q128as-info.log](https://github.com/openwrt/openwrt/files/14337893/bohong-bh25q128as-info.log)

Ansuel commented:
@themaverickdm v2 sent... btw i'm pushing this upstream just to permit us to include this in backports and forget about it. The backport patch will be much different since they changed lots of thing in recent kernel with the flash_info table.

Anyway thanks a lot for the dumps!

themaverickdm commented:
@Ansuel, thanks for taking this on, appreciated! And yes, I did find your patches on the kernel mailing list.

Any idea when this change might be available in openwrt snapshot builds?

Ansuel commented:
As soon as it's gets approved I will take care of backporting.

Ansuel commented:
@themaverickdm upstream they suggest to use compatible = "jedec,spi-nor"; and this patch removed... can you test this and do the same tests?

themaverickdm commented:
@Ansuel, I already have. Both the 5.15.y and 6.1.y kernels do not recognise the flash part with the following messages (below messages from 6.1.y):
```
[    0.726451] spi spi0.0: setup: ignoring unsupported mode bits a00
[    0.732850] spi-nor spi0.0: unrecognized JEDEC id bytes: 68 40 18 68 40 18
[    0.739725] spi-nor: probe of spi0.0 failed with error -2
```

Ansuel commented:
@themaverickdm i found the patch needed... would be ideal if you can test it https://github.com/openwrt/openwrt/pull/14738 for both 5.15 and 6.1 and compare the info reported with those command if they correctly match.

themaverickdm commented:
> @themaverickdm i found the patch needed... would be ideal if you can test it #14738 for both 5.15 and 6.1 and compare the info reported with those command if they correctly match.

I will try this out when I get a chance, and let you know. Thanks!

themaverickdm commented:
> @themaverickdm i found the patch needed... would be ideal if you can test it #14738 for both 5.15 and 6.1 and compare the info reported with those command if they correctly match.

@Ansuel, sorry for the delayed response. Please find attached a small log file with the result of initial testing of the generic spi-nor driver. Everything except the "manufacturer" works as expected.

[bh25q128as-generic-spi-nor-test.log](https://github.com/openwrt/openwrt/files/14497683/bh25q128as-generic-spi-nor-test.log)

Ansuel commented:
@themaverickdm can you change this to drop patch for 6.1?

themaverickdm committed:
generic: 5.15: Make support for BoHong bh25q128as generic

Move the patch for BoHong bh25q128as out of ramips to make it
generic. Not including 6.1.y since the mtd subsystem has changed,
and does not need these changes.

Patch was dropped with ramips updating to 6.1, hence we reintroudce it
here for 5.15 generic.

5.15.y functionality was verified on a Wavlink WL-WN586X3 Rev.a.

Signed-off-by: R Maru <deviantmaru@gmail.com>
[ rebase and add extra info in commit description ]
Signed-off-by: Christian Marangi <ansuelsmth@gmail.com>

themaverickdm commented:
> @themaverickdm can you change this to drop patch for 6.1? (leave 5.15 alone as it's not affected by your mediatek target)

@Ansuel, done, please review.

themaverickdm commented:
@Ansuel, how does one do multi-target "make target/linux/refresh"? It looks like ath79 and pistachio need the refresh. I can do the individual targets one after the other, but that is tedious, and possibly error-prone. Thanks.

Ansuel commented:
I totally missed the state of this @themaverickdm can you help ?

themaverickdm commented:
@Ansuel, thanks for getting this through. I did not get time to work on this over the last month.
